lighthouse-core/lib/csp-evaluator.js

```diff
@@ -0,0 +1,172 @@
+  deprecatedDisownOpener: 'disown-opener is deprecated since CSP3. ' +
+    'Please, use the Cross-Origin-Opener-Policy header instead.',
+};
+
+const str_ = i18n.createMessageInstanceIdFn(__filename, UIStrings);
+
+/** @type {Record<number, string|LH.IcuMessage|Record<string, LH.IcuMessage>>} */
+const FINDING_TO_UI_STRING = {
+  [Type.MISSING_SEMICOLON]: UIStrings.missingSemicolon,
+  [Type.UNKNOWN_DIRECTIVE]: str_(UIStrings.unknownDirective),
+  [Type.INVALID_KEYWORD]: UIStrings.unknownKeyword,
+  [Type.MISSING_DIRECTIVES]: {
+    [Directive.BASE_URI]: str_(UIStrings.missingBaseUri),
+    [Directive.SCRIPT_SRC]: str_(UIStrings.missingScriptSrc),
+    [Directive.OBJECT_SRC]: str_(UIStrings.missingObjectSrc),
+  },
+  [Type.SCRIPT_UNSAFE_INLINE]: str_(UIStrings.unsafeInline),
+  [Type.NONCE_LENGTH]: str_(UIStrings.nonceLength),
+  [Type.NONCE_CHARSET]: str_(UIStrings.nonceCharset),
+  [Type.DEPRECATED_DIRECTIVE]: {
+    [Directive.REFLECTED_XSS]: str_(UIStrings.deprecatedReflectedXSS),
+    [Directive.REFERRER]: str_(UIStrings.deprecatedReferrer),
+    [Directive.DISOWN_OPENER]: str_(UIStrings.deprecatedDisownOpener),
+  },
+  [Type.STRICT_DYNAMIC]: str_(UIStrings.strictDynamic),
+  [Type.UNSAFE_INLINE_FALLBACK]: str_(UIStrings.unsafeInlineFallback),
+  [Type.ALLOWLIST_FALLBACK]: str_(UIStrings.allowlistFallback),
+  [Type.REPORTING_DESTINATION_MISSING]: str_(UIStrings.reportingDestinationMissing),
+  [Type.REPORT_TO_ONLY]: str_(UIStrings.reportToOnly),
+};
+
+/**
+ * @param {Finding} finding
+ * @return {LH.IcuMessage|string}
+ */
+function getTranslatedDescription(finding) {
+  let result = FINDING_TO_UI_STRING[finding.type];
+  if (!result) {
+    log.warn('CSP Evaluator', `No translation found for description: ${finding.description}`);
+    return finding.description;
+  }
+
+  // Return if translated result found.
+  if (i18n.isIcuMessage(result)) return result;
+
+  // If result was not translated, that means `finding.value` is included in the UI string.
+  if (typeof result === 'string') return str_(result, {keyword: finding.value || ''});
+
+  // Result is a record object, UI string depends on the directive.
+  result = result[finding.directive];
+  if (!result) {
+    log.warn('CSP Evaluator', `No translation found for description: ${finding.description}`);
+    return finding.description;
+  }
+
+  return result;
+}
+
+/**
+ * @param {Array<string>} rawCsps
+ * @return {Array<Finding>}
+ */
+function evaluateRawCspForFailures(rawCsps) {
+  return evaluateForFailure(rawCsps.map(c => new Parser(c).csp));
+}
+
+/**
+ * @param {Array<string>} rawCsps
+ * @return {Array<Finding>}
+ */
+function evaluateRawCspForWarnings(rawCsps) {
+  return evaluateForWarnings(rawCsps.map(c => new Parser(c).csp));
+}
+
+/**
+ * @param {Array<string>} rawCsps
+ * @return {Array<Array<Finding>>}
```

Comment:
Do what's in your heart, @adamraine!

Reply:
I don't think I understand the difference between the two. Is it just style?

We use `Finding[][]` here, but if the type doesn't have a name we should use `Array<{a: string, b: number}>`?

Reply:
> I don't think I understand the difference between the two. Is it just style?
> 
> We use `Finding[][]` here, but if the type doesn't have a name we should use `Array<{a: string, b: number}>`?

basically we couldn't all agree on a rule so it falls in our ambiguous style zone of leaving the choice up to the author, unless editing a file with strong existing usage one way or the other, at which point you should probably match the existing style.

I'm a fan of `Array<>` always because I like keeping it completely separate from tuples, I like having it unambiguous up front that it's an array of stuff instead of revising mental parsing at the very end of the type, it matches the actual implementation (`Array<>` is parameterized by its contents), and I was indoctrinated long ago by Closure's style and it's hard to unbrainwash that :)

I believe (though correct me if I'm wrong) that @patrickhulce and @connorjclark are fans of basically what you describe, which roughly lines up with tslint's [`array-type` `array-simple`](https://palantir.github.io/tslint/rules/array-type/) mode ("enforces use of `T[]` if `T` is a simple type (primitive or type reference)") though "simple" may include more things than that for them.

Reply:
In this PR, this is very tough, in this PR I'm going to take my talents to `Finding[][]`.